common/compression/stuffit.cpp

```diff
@@ -140,37 +142,47 @@ bool StuffItArchive::open(Common::SeekableReadStream *stream) {
 
 	_stream->skip(7); // unknown
 
+	Common::CRC16 crc;
+
 	while (_stream->pos() < _stream->size() && !_stream->eos() && _stream->pos() < archiveSize) {
-		byte resForkCompression = _stream->readByte();
-		byte dataForkCompression = _stream->readByte();
+		byte header[112];
+		_stream->read(header, sizeof(header));
+		Common::MemoryReadStream headStream(header, sizeof(header));
+		byte resForkCompression = headStream.readByte();
+		byte dataForkCompression = headStream.readByte();
 
-		byte fileNameLength = _stream->readByte();
+		byte fileNameLength = headStream.readByte();
 		Common::String name;
 
 		if (fileNameLength > 63)
 			error("File name length too long in stuffit archive: %d at 0x%x", fileNameLength, (int) (_stream->pos() - 3));
 
 		for (byte i = 0; i < fileNameLength; i++)
-			name += (char)_stream->readByte();
+			name += (char)headStream.readByte();
 
 		// Skip remaining bytes
-		_stream->skip(63 - fileNameLength);
+		headStream.skip(63 - fileNameLength);
 
 		MacFinderInfo finfo;
 
-		_stream->read(finfo.type, 4);
-		_stream->read(finfo.creator, 4);
-		finfo.flags = _stream->readUint16BE();
-		/* uint32 creationDate = */ _stream->readUint32BE();
-		/* uint32 modificationDate = */ _stream->readUint32BE();
-		uint32 resForkUncompressedSize = _stream->readUint32BE();
-		uint32 dataForkUncompressedSize = _stream->readUint32BE();
-		uint32 resForkCompressedSize = _stream->readUint32BE();
-		uint32 dataForkCompressedSize = _stream->readUint32BE();
-		/* uint16 resForkCRC = */ _stream->readUint16BE();
-		/* uint16 dataForkCRC = */ _stream->readUint16BE();
-		_stream->skip(6); // unknown
-		/* uint16 headerCRC = */ _stream->readUint16BE();
+		headStream.read(finfo.type, 4);
+		headStream.read(finfo.creator, 4);
+		finfo.flags = headStream.readUint16BE();
+		/* uint32 creationDate = */ headStream.readUint32BE();
+		/* uint32 modificationDate = */ headStream.readUint32BE();
+		uint32 resForkUncompressedSize = headStream.readUint32BE();
+		uint32 dataForkUncompressedSize = headStream.readUint32BE();
+		uint32 resForkCompressedSize = headStream.readUint32BE();
+		uint32 dataForkCompressedSize = headStream.readUint32BE();
+		uint16 resForkCRC = headStream.readUint16BE();
+		uint16 dataForkCRC = headStream.readUint16BE();
+		headStream.skip(6); // unknown
+		uint16 headerCRC = headStream.readUint16BE();
+
+		uint16 actualHeaderCRC = crc.crcFast(header, sizeof(header) - 2);
+
+		if (actualHeaderCRC != headerCRC)
+			error ("Header CRC mismatch: %04x vs %04x", actualHeaderCRC, headerCRC);
```

Comment:
May it be possible to add here `StuffItArchive::open(): ` prefix and also the name of the broken file? Same for the `error()` below